added functions to work with OH rest/links/ endpoint

snpz opened the pull request:
Hi @spacemanspiff2007 

Here is my first try for a pull request.

A few notes regarding the changes:
- i changed the parameter name from thing_name to thing_uid as this is what is actually used (we could of course change this to the thing_name and look the UID up via async_get_thing before accessing the links-endpoint)
- the thing UID is in a format like AAAAAA:BBBBBB:CCCCCCC:DDDDD which needs to be URL-encoded to be able to work with the api (I'm doing this using urllib.parse.quote(url)

Before I found the missing encoding to be the issue I always got the OpenHabNotReadyYetException from _check_http_response. This occurs because you say that OpenHAB is not ready when there is a http-response-status 500 which means InternalServerError but could actually be anything. 

Also I'm not sure about error handling and exceptions and what to propagate to the caller and what not. On any errors except disconnection the functions currently would return false without any possibility for the caller to get the error besides looking in the logs.

snpz commented on HABApp/openhab/http_connection.py:
```diff
@@ -320,6 +321,47 @@ async def async_get_item(self, item_name: str) -> dict:
+        channelUID = urllib.parse.quote(thing_UID + ":" + channel)
+        url = self.__get_openhab_url('rest/links/{item:s}/{channel:s}',item=item_name,channel=channelUID)
+        
+        fut = self.__session.put(url,json={})
```
I also found this strange, but it gets the information from the url-path and if a payload exists, compares them with the payload and throws an error if they don't match.

See the [relevant code in the rest api](https://github.com/eclipse/smarthome/blob/e25f1bd548126ef44fbdc90dd1dd0f85e83bdef0/bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/internal/link/ItemChannelLinkResource.java#L116).

After having a second look - the difference is you can't use the configuration parameter this way. I wasn't aware there are link-configurations at all, but it seems like there are different profiles one could use for setting up a link. The available profiles could be queried on the /profile-types endpoint. This seems to be an advanced feature though I personally never needed.

We could add an optional profile parameter (defaults to "system:default" and verify the given profile against the profile-types endpoint but this would not be enough if the configuration is extended in the future for example.

How do you want to support those, if at all?

spacemanspiff2007 replied:
Can the payload be omitted? Then I'd go for that.

snpz replied:
It can be omitted yes, that is how it currently works.

But as said [here](https://github.com/spacemanspiff2007/HABApp/issues/150) we probably should allow passing the configuration when adding a link as Dict<str,Any>. 
I found posts online where people are forced to use the link configuration if e.g. specific bindings are involved. The configuration is very generic with Dict<str,Any> though. Not sure how this can be serialized/deserialized.
I think it would be the best move to make no assumptions of the configuration content at all, just pass it through and let the OpenHAB Api handle errors related to non-existing profiles or wrong configurations.

To make use of the new definition I'd also add a thing_get_link() to be able to access the configuration.

snpz commented on HABApp/openhab/http_connection.py:
```diff
@@ -320,6 +321,47 @@ async def async_get_item(self, item_name: str) -> dict:
+            return False
+
+        # check for item existence first, otherwhise OpenHAB creates a new item when we add a link with an unknown itemname
+        exists = await self.async_item_exists(item_name)        
+        if not exists:
+            return False
```
I followed the inheritance and this call would use 
``` python
item = HABApp.core.Items.get_item(name)
```
from the ```BaseItem```. But, if I understand correctly, this also includes non-OpenHAB items which you can't link to in OpenHAB of course. Am I missing something?

spacemanspiff2007 replied:
The assertion after the ensures that it is only an OpenhabItem
```python
        item = HABApp.core.Items.get_item(name)
        assert isinstance(item, cls), f'{cls} != {type(item)}'
```

snpz replied:
I should have looked up the cls keyword before, no I get how it works.

But: I still don't understand why you would want to introduce this dependency to the http_connection.
Wouldn't this also introduce an issue when I create an Item in OpenHAB and a link for the Item directly afterwards? As far as I understand HABApp the internal ItemRegistry used in your proposal is asynchronously updated based on events from OpenHAB. So if the event was not handled yet, ```HABApp.core.Items.get_item(name)``` would raise an exception despite the Item existing in OpenHAB.

If you insist I can change it of course, but I would like to understand your reasoning - also to be more in line with your thoughts in future Pull Requests.

spacemanspiff2007 replied:
> Wouldn't this also introduce an issue when I create an Item in OpenHAB and a link for the Item directly afterwards?

You are right, I was more thinking about linking existing items. With the additional check we are on the safe side.

snpz commented:
I have another question regarding the type hints: (as I said I'm pretty new to Python)
Why would you add them to http_connection which is only used internally but not to oh_interface which is exposed to the end-user?

Maybe I don't understand them yet coming from compiled languages where I would avoid a generic object type and type-checking as much as I could.

NickWaterton commented:
Stefan,

It’s just a coding style. Most Python programmers don’t use type-checking, hints etc. -  because it’s so much easier (and they are not required in Python). Other people coming from C-type languages feel lost without them. I haven’t seen anyone else’s code quite so thorough as spaceman’s though. It is unique.

Nick Waterton P.Eng.

Unsubscribe from GE's commercial electronic messages: http://sc.ge.com/*casl-unsubscribe 
Désabonner des messages électroniques commerciaux de GE: http://sc.ge.com/*lcap-desabonnement

snpz commented:
Nick, thanks for the explanation. I'm one of those feeling lost without types, so I'm happy if it's used. :) It can't be that bad having the ability to avoid runtime errors. ;)

(I deleted the post because I saw that type hints are also in oh_interface which made my question obsolete)

NickWaterton commented:
Well this is a bit of a philosophical discussion, but you don't get runtime errors like you do in a complied language. Python takes care of most things automatically, and will show you the exact error if you try to add an integer to a string or something. It's a completely different paradigm.

snpz commented on HABApp/openhab/http_connection.py:
```diff
@@ -320,6 +321,47 @@ async def async_get_item(self, item_name: str) -> dict:
+
+    async def async_link_exists(self, item_name, thing_UID, channel) -> bool:
+        # rest/links/ endpoint needs the channel to be url encoded (AAAA:BBBB:CCCC:0#NAME -> AAAA%3ABBBB%3ACCCC%3A0%23NAME)
+        channelUID = urllib.parse.quote(thing_UID + ":" + channel)
```
done in c9821e0

snpz commented on HABApp/openhab/http_connection.py:
```diff
@@ -320,6 +321,47 @@ async def async_get_item(self, item_name: str) -> dict:
+        if self.config.general.listen_only:
+            return False
+
+        # rest/links/ endpoint needs the channel to be url encoded (AAAA:BBBB:CCCC:0#NAME -> AAAA%3ABBBB%3ACCCC%3A0%23NAME)
```
done in c9821e0

snpz commented on HABApp/openhab/http_connection.py:
```diff
@@ -320,6 +321,47 @@ async def async_get_item(self, item_name: str) -> dict:
+            return False
+        
+        # rest/links/ endpoint needs the channel to be url encoded (AAAA:BBBB:CCCC:0#NAME -> AAAA%3ABBBB%3ACCCC%3A0%23NAME)
+        channelUID = urllib.parse.quote(thing_UID + ":" + channel)
```
done in c9821e0

spacemanspiff2007 commented on HABApp/openhab/http_connection.py:
```diff
@@ -320,6 +321,53 @@ async def async_get_item(self, item_name: str) -> dict:
+        # rest/links/ endpoint needs the channel to be url encoded 
+        # (AAAA:BBBB:CCCC:0#NAME -> AAAA%3ABBBB%3ACCCC%3A0%23NAME)
+        # otherwise the REST-api returns HTTP-Status 500 InternalServerError
+        channel_uid = urllib.parse.quote(f"{thing_uid}:{channel}")
+        url = self.__get_openhab_url('rest/links/{item:s}/{channel:s}', item=item_name, channel=channel_uid)
```
Imho we should move quoting into the function

```python
    def __get_openhab_url(self, url: str, *args, **kwargs) -> str:
        assert not url.startswith('/')

        url = url.format(*args, **kwargs)
        if url.startswith('rest/links/'):
            # rest/links/ endpoint needs the channel to be url encoded 
            # (AAAA:BBBB:CCCC:0#NAME -> AAAA%3ABBBB%3ACCCC%3A0%23NAME)
            # otherwise the REST-api returns HTTP-Status 500 InternalServerError
            url = urllib.parse.quote(url)
        return f'http://{self.__host:s}:{self.__port:d}/{url:s}'
```

snpz replied:
done in d5a9486

snpz commented on HABApp/openhab/http_connection.py:
```diff
@@ -320,6 +321,53 @@ async def async_get_item(self, item_name: str) -> dict:
+        # check for item existence first, otherwhise OpenHAB creates a new item when we add a link with an unknown itemname
+        exists = await self.async_item_exists(item_name)        
+        if not exists:
+            return False
```
done in d5a9486

snpz replied:
@spacemanspiff2007 
the code in async_get_item still uses OpenhabItemNotFoundError which seems to be used nowhere else besides a test. Should I replace this with ItemNotFoundException and remove the OpenhabItemNotFoundError completely?

```
    async def async_get_item(self, item_name: str) -> dict:
        fut = self.__session.get(self.__get_openhab_url('rest/items/{:s}', item_name))
        ret = await self._check_http_response(fut, accept_404=True)
        if ret.status == 404:
            raise HABApp.openhab.exceptions.OpenhabItemNotFoundError(f'Item {item_name} not found!')
        if ret.status >= 300:
            return {}
        else:
            return await ret.json(encoding='utf-8')
```

spacemanspiff2007 commented on HABApp/openhab/http_connection.py:
```diff
@@ -325,22 +333,16 @@ async def async_thing_remove_link(self, thing_uid: str, channel: str, item_name:
-        # otherwise the REST-api returns HTTP-Status 500 InternalServerError
-        channel_uid = urllib.parse.quote(f"{thing_uid}:{channel}")
+        channel_uid = f"{thing_uid}:{channel}"
         url = self.__get_openhab_url('rest/links/{item:s}/{channel:s}', item=item_name, channel=channel_uid)
```
```suggestion
        url = self.__get_openhab_url('rest/links/{item:s}/{thing_uid:s}:{channel:s}', item=item_name, thing_uid=thing_uid, channel=channel_uid)
```

snpz replied:
I added a self.__get_link_url() method to have it in one place. This also gets rid of the check in __get_openhab_url.

spacemanspiff2007 commented on HABApp/openhab/http_connection.py:
```diff
@@ -355,12 +357,9 @@ async def async_thing_add_link(self, thing_uid: str, channel: str, item_name: st
-        # otherwise the REST-api returns HTTP-Status 500 InternalServerError
-        channel_uid = urllib.parse.quote(f"{thing_uid}:{channel}")
+        channel_uid = f"{thing_uid}:{channel}"
         url = self.__get_openhab_url('rest/links/{item:s}/{channel:s}', item=item_name, channel=channel_uid)
```
```suggestion
        url = self.__get_openhab_url('rest/links/{item:s}/{thing_uid:s}:{channel:s}', item=item_name, thing_uid=thing_uid, channel=channel_uid)
```

snpz replied:
I added a self.__get_link_url() method to have it in one place. This also gets rid of the check in __get_openhab_url.

spacemanspiff2007 commented on HABApp/openhab/http_connection.py:
```diff
@@ -325,22 +333,16 @@ async def async_thing_remove_link(self, thing_uid: str, channel: str, item_name:
-        channel_uid = urllib.parse.quote(f"{thing_uid}:{channel}")
+    async def async_thing_link_exists(self, thing_uid: str, channel: str, item_name: str) -> bool:        
+        channel_uid = f"{thing_uid}:{channel}"
         url = self.__get_openhab_url('rest/links/{item:s}/{channel:s}', item=item_name, channel=channel_uid)
```
```suggestion
        url = self.__get_openhab_url('rest/links/{item:s}/{thing_uid:s}:{channel:s}', item=item_name, thing_uid=thing_uid, channel=channel_uid)
```

snpz replied:
I added a self.__get_link_url() method to have it in one place. This also gets rid of the check in __get_openhab_url.

spacemanspiff2007 commented on HABApp/openhab/http_connection.py:
```diff
@@ -320,6 +329,44 @@ async def async_get_item(self, item_name: str) -> dict:
+        ret = await self._check_http_response(fut, accept_404=True)
+        return ret.status == 200
+
+    async def async_thing_add_link(self, thing_uid: str, channel: str, item_name: str) -> bool:
+        if self.config.general.listen_only:
+            return False
+
+        # check for item existence first, otherwhise OpenHAB creates a new item when we add a link with an unknown itemname
+        exists = await self.async_item_exists(item_name)        
+        if not exists:
+            raise ItemNotFoundException(f'Item "{item_name}" does not exist')
+        
+        channel_uid = f"{thing_uid}:{channel}"
+        url = self.__get_openhab_url('rest/links/{item:s}/{channel:s}', item=item_name, channel=channel_uid)
+        
+        fut = self.__session.put(url, json={})
+
+        ret = await self._check_http_response(fut)
+        return ret.status == 200
```
```suggestion
    async def async_thing_add_link(self, thing_uid: str, channel: str, item_name: str, config: typing.Optional[ItemChannelLinkDefinition]) -> bool:
        if self.config.general.listen_only:
            return False

        # check for item existence first, otherwhise OpenHAB creates a new item when we add a link with an unknown itemname
        exists = await self.async_item_exists(item_name)        
        if not exists:
            raise ItemNotFoundException(f'Item "{item_name}" does not exist')
        
        url = self.__get_openhab_url('rest/links/{item:s}/{thing_uid:s}:{channel:s}', item=item_name, thing_uid=thing_uid, channel=channel_uid)
        
        fut = self.__session.put(url, json=o.dict(by_alias=True) if isinstance(config, ItemChannelLinkDefinition) else None)

        ret = await self._check_http_response(fut)
        return ret.status == 200
```

spacemanspiff2007 replied:
This would allow passing in ItemChannelLinkDefinition objects

snpz replied:
When using the ItemChannelLinkDefinition we could also use it as a parameter directly and remove item_name, thing_name and channel because item_name and channel_uid are properties of the definition.

To me the usage of the definitions is incosistent currently. Some methods in http_connection return the parsed item definition directly, others a dict which is parsed into the definition in oh_interface. One could also use the OpenhabItemDefinition as a parameter in async_create_item instead of the long parameter list.

Is it possible to add methods to such a definition class?

snpz replied:
I found a [good description of the configuration](https://github.com/eclipse/smarthome/blob/11764d8ae824cadc95de5dd228f9f077f4859dcc/docs/documentation/development/bindings/profiles.md).
```
Number <itemName> { channel="<bindingID>:<thing-typeID>:<thingName>:<channelName>"[profile="offset", offset="<value>"]}
```
the ```[profile="offset", offset="<value>"]``` describes the configuration meaning there is a key for the profile and as many other keys as needed to configure the profile.
I'm not sure if it is necessary to model this as class or just have the dict<str,Any> available if needed.

spacemanspiff2007 replied:
> When using the ItemChannelLinkDefinition we could also use it as a parameter directly and remove item_name, thing_name and channel because item_name and channel_uid are properties of the definition.

Sounds good

> Is it possible to add methods to such a definition class?

Sure, but let's finish one PR at a time. I'm excited that someone is collaborating but unfortunately I have not very much experience using the review tools from github.

spacemanspiff2007 commented on HABApp/openhab/http_connection.py:
```diff
@@ -320,6 +331,35 @@ async def async_get_item(self, item_name: str) -> dict:
+        if not exists:
+            raise ItemNotFoundException(f'Item "{item_name}" does not exist')
+        
+        fut = self.__session.put(self.__get_link_url(thing_uid, channel, item_name), json={})
```
Is the parameter json really required?
```suggestion
        fut = self.__session.put(self.__get_link_url(thing_uid, channel, item_name))
```

spacemanspiff2007 commented on HABApp/openhab/http_connection.py:
```diff
@@ -320,6 +331,35 @@ async def async_get_item(self, item_name: str) -> dict:
+        if self.config.general.listen_only:
+            return False
+
+        fut = self.__session.delete(self.__get_link_url(thing_uid, channel, item_name), json={})
```
Is the parameter json really required?
```suggestion
        fut = self.__session.delete(self.__get_link_url(thing_uid, channel, item_name))
```

snpz replied:
nope, removed

snpz commented:
@spacemanspiff2007 
I added your requests and also the use of ItemChannelLinkDefinition. This is how it would work from the user perspective:
``` python
        item = "MyTest"
        channel = "homematic:HG-BC-RT-TRX-CyN:7d4f981f:MKF0060051:0#SIGNAL_STRENGTH"

        ohi = self.openhab

        exists = ohi.thing_link_exists(channel,item)
        print(f"link exists: {exists}")

        exists = ohi.item_exists(item)
        print(f"item exists: {exists}")
        if exists:
            removed = ohi.remove_item(item)
            print("item removed: {removed}")

        try:
            created = ohi.create_item("Number:Intensity",item,item)
            print(f"item created: {created}")
        except Exception as e:
             print(f"Exception: {e}")

        link = ItemChannelLinkDefinition(channelUID=channel,itemName="non-existing-item",configuration={"profile":"system:offset","offset":7})

        try:
            addedInvalid = ohi.thing_add_link(link)
            print(f"link to non-existing item added: {addedInvalid}")
        except ItemNotFoundException as e:
            print(e)

        link.item_name = item
        addedCorrect = ohi.thing_add_link(link)
        print(f"correct link added: {addedCorrect}")

        try:
            ohi.thing_get_link(channel,"non-existing-item")
        except LinkNotFoundError as e:
            print(e)

        createdLink = ohi.thing_get_link(channel,item)
        print(f"retrieved link:")
        pprint.pprint(createdLink)

        exists = ohi.thing_link_exists(channel,item)
        print(f"link exists: {exists}")

        removed = ohi.thing_remove_link(channel,item)
        print(f"link removed: {removed}")

        exists = ohi.thing_link_exists(channel,item)
        print(f"link exists: {exists}")
```

Output is
```
link exists: False
item exists: False
item created: True
Item "non-existing-item" does not exist
correct link added: True
Link non-existing-item -> homematic:HG-BC-RT-TRX-CyN:7d4f981f:MKF0060051:0#SIGNAL_STRENGTH not found!
retrieved link:
ItemChannelLinkDefinition(item_name='MyTest', channel_uid='homematic:HG-BC-RT-TRX-CyN:7d4f981f:MKF0060051:0#SIGNAL_STRENGTH', configuration={'offset': 7.0, 'profile': 'system:offset'})
link exists: True
link removed: True
link exists: False
```

I changed all calls to use the channel_uid directly as the Api does instead of separating thing_uid and channel. I also added the LinkNotFoundError.

I've not yet pushed the changes because I wanted to make sure that it looks okay to you first.
And...yeah. I also would like to finish my first PR. :)

snpz commented on HABApp/openhab/http_connection.py:
```diff
@@ -320,6 +331,35 @@ async def async_get_item(self, item_name: str) -> dict:
+        return ret.status == 200
+
+    async def async_thing_link_exists(self, thing_uid: str, channel: str, item_name: str) -> bool:        
+        fut = self.__session.get(self.__get_link_url(thing_uid, channel, item_name), json={})
```
nope, removed

spacemanspiff2007 commented:
> I changed all calls to use the channel_uid directly as the Api does instead of separating thing_uid and channel. I also added the LinkNotFoundError.

Looks good, but then we should remove the ``thing_`` prefix again.

```ItemChannelLinkDefinition(channelUID=channel,itemName="non-existing-item",...)```
The user should have to specify ```channel_uid```, I'll take a look at the validater, maybe I've gotta inherit from a different base class to fix this.

There is a ``conf_testing`` folder in the repo. I use it to run tests again a running openhab test instance.
The Bindings Astro and Expire are installed there. You can create a new file "test_interface_links" and add the above snippets there (maybe split it in different functions).

The last thing that is then missing is the docstring (the strings with """) directly under the function definition in oh_interface.

snpz commented:
> `ItemChannelLinkDefinition(channelUID=channel,itemName="non-existing-item",...)`
> The user should have to specify `channel_uid`, I'll take a look at the validater, maybe I've gotta inherit from a different base class to fix this.
I found this strange too. I also had to use ```link.json(by_alias=True)``` to get the aliases in the created Json instead of the property names. 
I think turning it around and name the property of ItemChannelLInkDefinition channelUID and alias it with channel_uid would behave as needed. It still is a bit confusing though.

snpz commented:
I found this issue describing the behaviour, maybe that helps decide which way to go: https://github.com/samuelcolvin/pydantic/issues/565

________________________________
From: spacemanspiff2007 <notifications@github.com>
Sent: Saturday, April 25, 2020 5:43:58 AM
To: spacemanspiff2007/HABApp <HABApp@noreply.github.com>
Cc: Steffen <s.petz@simpligo.org>; Author <author@noreply.github.com>
Subject: Re: [spacemanspiff2007/HABApp] added functions to work with OH rest/links/ endpoint (#149)


I changed all calls to use the channel_uid directly as the Api does instead of separating thing_uid and channel. I also added the LinkNotFoundError.

Looks good, but then we should remove the thing_ prefix again.

ItemChannelLinkDefinition(channelUID=channel,itemName="non-existing-item",...)
The user should have to specify channel_uid, I'll take a look at the validater, maybe I've gotta inherit from a different base class to fix this.

There is a conf_testing folder in the repo. I use it to run tests again a running openhab test instance.
The Bindings Astro and Expire are installed there. You can create a new file "test_interface_links" and add the above snippets there (maybe split it in different functions).

The last thing that is then missing is the docstring (the strings with """) directly under the function definition in oh_interface.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/spacemanspiff2007/HABApp/pull/149#issuecomment-619314116>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AHIQ5TQ7QE2CZZEUYVFI4DDROJL75ANCNFSM4MNM3XZQ>.

snpz commented:
> There is a `conf_testing` folder in the repo. I use it to run tests again a running openhab test instance.
> The Bindings Astro and Expire are installed there. You can create a new file "test_interface_links" and add the above snippets there (maybe split it in different functions).

I'll have a look.

spacemanspiff2007 commented:
> two items aren't created during the tests but instead seem to exist in your installation (TestString, TestGroupAVG)

I really should move the creation into HABApp.

> When I start writing the tests, what would I need to do to execute only my file instead of having the whole suite run on startup (incl. the benchmark)?

Just rename the existing tests to e.g. ``.py_`` or move them out of the rules folder.

spacemanspiff2007 commented on conf_testing/rules/test_openhab_interface_links.py:
```diff
@@ -0,0 +1,97 @@
+
+        self.openhab.remove_item(self.item_name)
+
+    def __find_astro_sun_thing(self) -> str:
+        found_uid: str = ""
+        for item in get_all_items():
+            if isinstance(item, Thing) and item.name.startswith("astro:sun"):
+                found_uid = item.name
+
+        return found_uid
```
This doesn't change, should we move it in ``def set_up(self):``?

snpz replied:
Item creation doesn't change either, so I moved all of them to set_up() and put the member declaration in ``` __init__ ``` additionally to avoid the linter warning for "member definition outside of __init__".

```
    astro_sun_thing: str = ""
```

spacemanspiff2007 commented on conf_testing/rules/test_openhab_interface_links.py:
```diff
@@ -0,0 +1,97 @@
+from conf_testing.lib.HABAppTests import ItemWaiter, OpenhabTmpItem, TestBaseRule, get_openhab_test_states, get_openhab_test_types
+
+class TestOpenhabInterfaceLinks(TestBaseRule):
+    astro_sun_thing: str
+    item_name: str
+    channel_uid: str
```
this can be moved into ``__init__``

snpz replied:
I wasn't even aware that my syntax meant a static member. Another thing learned.

spacemanspiff2007 replied:
Yea, that's how I basically learned my whole python knowledge ;-)

spacemanspiff2007 commented on conf_testing/rules/test_openhab_interface_links.py:
```diff
@@ -0,0 +1,97 @@
+        changed_def.configuration["profile"] = "system:offset"
+        changed_def.configuration["offset"] = 7.0
+
+        changed_created = self.oh.create_link(changed_def)
+        assert changed_created
```
```suggestion
        assert self.oh.create_link(changed_def)
```

snpz commented on conf_testing/rules/test_openhab_interface_links.py:
```diff
@@ -0,0 +1,97 @@
+    def test_remove_link(self):
+        created = self.oh.create_link(self.__get_link_def())
+        assert created
+
```
I kept the creation there to not have to rely on test execution order and to be able to execute the test function independently. Hope that's ok.

snpz commented:
> 
> 
> Nice work!

Thanks. :)

spacemanspiff2007 commented:
Could you do a git rebase? Then this can be merged.

snpz commented:
> 
> 
> Could you do a git rebase? Then this can be merged.

Hi, I did a ```git rebase upstream/Develop```

spacemanspiff2007 commented:
Congrats

snpz commented:
time for the next one! ;)

snpz commented:
@spacemanspiff2007 
Any insights on this yet? I stumbled over it again and it should be solved before going to master I think.

> 
> 
> > `ItemChannelLinkDefinition(channelUID=channel,itemName="non-existing-item",...)`
> > The user should have to specify `channel_uid`, I'll take a look at the validater, maybe I've gotta > inherit from a different base class to fix this.
> 
> I found this strange too. I also had to use `link.json(by_alias=True)` to get the aliases in the created Json instead of the property names.
> I think turning it around and name the property of ItemChannelLInkDefinition channelUID and alias it with channel_uid would behave as needed. It still is a bit confusing though.

spacemanspiff2007 commented:
> I found this strange too. I also had to use link.json(by_alias=True) to get the aliases in the created Json instead of the property names.
I think turning it around and name the property of ItemChannelLInkDefinition channelUID and alias it with channel_uid would behave as needed. It still is a bit confusing though.

Maybe we revert to the function with the three properties instead of the class again.
That way the user doesn't have to deal with it.

snpz commented:
> 
> 
> > I found this strange too. I also had to use link.json(by_alias=True) to get the aliases in the created Json instead of the property names.
> > I think turning it around and name the property of ItemChannelLInkDefinition channelUID and alias it with channel_uid would behave as needed. It still is a bit confusing though.
> 
> Maybe we revert to the function with the three properties instead of the class again.
> That way the user doesn't have to deal with it.

I just took another look to be sure:
If we use the config setting ``` allow_population_by_fieldname ```
``` python
class ItemChannelLinkDefinition(BaseModel):
    class Config:
        allow_population_by_field_name = True

    item_name: str = Field(alias='itemName')
    channel_uid: str = Field(alias='channelUID')
    configuration: Dict[str, Any] = {}
```
see this test which passes
```python
    def test_link_def(self):
        link_def = self.__get_link_def()
        assert link_def.item_name == self.item_name
        assert link_def.channel_uid == self.channel_uid

        link_def = ItemChannelLinkDefinition(channel_uid=self.channel_uid, item_name=self.item_name)
        link_def_alternate = ItemChannelLinkDefinition(channelUID=self.channel_uid, itemName=self.item_name)

        thrown = False
        try:
            link_def.itemName = "test"
        except ValueError as att: #ValueError: "ItemChannelLinkDefinition" object has no field "itemName"
            thrown = True
        
        assert thrown
        assert link_def == link_def_alternate
```

I think this could hide it from the user.

In http_connection, where link_def.dict() is used, I removed the by_alias=True to see what happens. This of course results in json having item_name instead of itemName and channel_uid instead of channelUID. Strangely enough it still works flawlessly - I also tried the calls directly in the swagger interface.

spacemanspiff2007 commented:
> Strangely enough it still works flawlessly - I also tried the calls directly in the swagger interface.

O_o
